File: src/main/java/com/uiptv/util/EpisodeTitleFormatter.java
package com.uiptv.util;

import java.text.Normalizer;
import java.util.List;
import java.util.Locale;

import static com.uiptv.util.StringUtils.isBlank;

public final class EpisodeTitleFormatter {
    private static final List<String> ORDINAL_URDU_PREFIXES = List.of(
            "pehla", "pehli", "doosra", "doosri", "teesra", "teesri",
            "chautha", "chauthi", "paanchva", "paanchvi", "paanchwa", "paanchwi",
            "chhata", "chhati", "saatva", "saatvi", "aathva", "aathvi",
            "nauva", "nauvi", "dasva", "dasvi"
    );
    private static final List<String> EPISODE_WORDS = List.of("qist", "kist", "episode", "ep", "kadi");

    private EpisodeTitleFormatter() {
    }

    public static String buildEpisodeDisplayTitle(String season, String episodeNumber, String title) {
        String normalizedSeason = isBlank(season) ? "1" : season;
        String seasonLabel = I18n.formatSeasonLabel(normalizedSeason);
        String episodeLabel = I18n.formatEpisodeLabel(episodeNumber);
        String cleanTitle = stripGenericEpisodeTitle(normalizedSeason, episodeNumber, title);
        if (isBlank(cleanTitle)) {
            return seasonLabel + " - " + episodeLabel;
        }
        return seasonLabel + " - " + episodeLabel + ": " + cleanTitle;
    }

    public static boolean isGenericEpisodeTitle(String title) {
        String value = safe(title).trim();
        if (isBlank(value)) {
            return true;
        }
        String normalized = normalizeAsciiWords(value);
        return matchesSimpleEpisodeNumber(normalized)
                || matchesOrdinalSeasonEpisode(normalized)
                || matchesSeasonEpisode(normalized);
    }

    static String stripGenericEpisodeTitle(String season, String episodeNumber, String title) {
        String value = safe(title).trim();
        if (isBlank(value)) {
            return "";
        }
        if (isGenericEpisodeTitle(value) || matchesLocalizedDisplayLabel(season, episodeNumber, value)) {
            return "";
        }
        return value;
    }

    static boolean matchesLocalizedDisplayLabel(String season, String episodeNumber, String title) {
        String normalizedTitle = normalizeForComparison(title);
        if (isBlank(normalizedTitle)) {
            return true;
        }
        String normalizedSeason = isBlank(season) ? "1" : season;
        String seasonLabel = I18n.formatSeasonLabel(normalizedSeason);
        String episodeLabel = I18n.formatEpisodeLabel(episodeNumber);
        String numericSeasonLabel = I18n.tr("autoSeasonNumber", I18n.formatNumber(normalizedSeason));
        String numericEpisodeLabel = I18n.tr("autoEpisodeNumber", I18n.formatNumber(isBlank(episodeNumber) ? "-" : episodeNumber));

        for (String candidate : List.of(
                seasonLabel + " - " + episodeLabel,
                seasonLabel + ": " + episodeLabel,
                seasonLabel + " " + episodeLabel,
                seasonLabel + "  " + episodeLabel,
                episodeLabel,
                seasonLabel + " - " + numericEpisodeLabel,
                seasonLabel + ": " + numericEpisodeLabel,
                seasonLabel + " " + numericEpisodeLabel,
                numericSeasonLabel + " - " + numericEpisodeLabel,
                numericSeasonLabel + ": " + numericEpisodeLabel,
                numericSeasonLabel + " " + numericEpisodeLabel,
                numericSeasonLabel + " - " + episodeLabel,
                numericSeasonLabel + ": " + episodeLabel,
                numericSeasonLabel + " " + episodeLabel,
                numericEpisodeLabel
        )) {
            if (normalizeForComparison(candidate).equals(normalizedTitle)) {
                return true;
            }
        }
        return false;
    }

    static String normalizeForComparison(String value) {
        String normalized = Normalizer.normalize(normalizeDigitsToAscii(safe(value)), Normalizer.Form.NFKC)
                .toLowerCase(I18n.getCurrentLocale());
        return normalized.replaceAll("[^\\p{L}\\p{N}]+", " ")
                .replaceAll("\\s+", " ")
                .trim();
    }

    private static String normalizeDigitsToAscii(String value) {
        StringBuilder out = new StringBuilder(value.length());
        for (int i = 0; i < value.length(); i++) {
            char ch = value.charAt(i);
            if (Character.isDigit(ch)) {
                int numeric = Character.getNumericValue(ch);
                if (numeric >= 0 && numeric <= 9) {
                    out.append((char) ('0' + numeric));
                    continue;
                }
            }
            out.append(ch);
        }
        return out.toString();
    }

    private static String safe(String value) {
        return value == null ? "" : value;
    }

    private static String normalizeAsciiWords(String value) {
        return normalizeDigitsToAscii(safe(value))
                .toLowerCase(Locale.ROOT)
                .replace('.', ' ')
                .replace(':', ' ')
                .replace('-', ' ')
                .trim()
                .replaceAll("\\s+", " ");
    }

    private static boolean matchesSimpleEpisodeNumber(String value) {
        if (value.matches("^episode\\s+\\d+$")) {
            return true;
        }
        if (value.matches("^ep\\s+\\d+$")) {
            return true;
        }
        return value.matches("^e\\d+$");
    }

    private static boolean matchesOrdinalSeasonEpisode(String value) {
        String[] parts = value.split(" ");
        if (parts.length != 4 || !"season".equals(parts[1]) || !isEpisodeWord(parts[2])) {
            return false;
        }
        return ORDINAL_URDU_PREFIXES.contains(parts[0]) && isDigits(parts[3]);
    }

    private static boolean matchesSeasonEpisode(String value) {
        String[] parts = value.split(" ");
        if (parts.length != 4 || !"season".equals(parts[0]) || !isEpisodeWord(parts[2])) {
            return false;
        }
        return isDigits(parts[1]) && isDigits(parts[3]);
    }

    private static boolean isEpisodeWord(String value) {
        return EPISODE_WORDS.contains(value);
    }

    private static boolean isDigits(String value) {
        return !isBlank(value) && value.chars().allMatch(Character::isDigit);
    }
}

File: server/src/main/kotlin/com/uiptv/util/EpisodeTitleFormatter.kt
package com.uiptv.util

import java.text.Normalizer
import java.util.Locale
import java.util.regex.Pattern

object EpisodeTitleFormatter {
    private val ORDINAL_URDU_PREFIXES = listOf(
        "pehla", "pehli", "doosra", "doosri", "teesra", "teesri",
        "chautha", "chauthi", "paanchva", "paanchvi", "paanchwa", "paanchwi",
        "chhata", "chhati", "saatva", "saatvi", "aathva", "aathvi",
        "nauva", "nauvi", "dasva", "dasvi"
    )
    private val EPISODE_WORDS = listOf("qist", "kist", "episode", "ep", "kadi")

    @JvmStatic
    fun buildEpisodeDisplayTitle(season: String?, episodeNumber: String?, title: String?): String {
        val normalizedSeason = if (StringUtils.isBlank(season)) "1" else season!!
        val seasonLabel = I18n.formatSeasonLabel(normalizedSeason)
        val episodeLabel = I18n.formatEpisodeLabel(episodeNumber)
        var cleanTitle = stripGenericEpisodeTitle(normalizedSeason, episodeNumber, title)
        cleanTitle = stripLeadingEpisodeMarker(normalizedSeason, episodeNumber, cleanTitle)
        return if (StringUtils.isBlank(cleanTitle)) {
            "$seasonLabel - $episodeLabel"
        } else {
            "$seasonLabel - $episodeLabel: $cleanTitle"
        }
    }

    @JvmStatic
    fun isGenericEpisodeTitle(title: String?): Boolean {
        val value = safe(title).trim()
        if (StringUtils.isBlank(value)) {
            return true
        }
        val normalized = normalizeAsciiWords(value)
        return matchesSimpleEpisodeNumber(normalized) ||
            matchesOrdinalSeasonEpisode(normalized) ||
            matchesSeasonEpisode(normalized) ||
            matchesLocalizedEpisodeWordTitle(normalized)
    }

    @JvmStatic
    fun stripGenericEpisodeTitle(season: String?, episodeNumber: String?, title: String?): String {
        val value = safe(title).trim()
        if (StringUtils.isBlank(value)) {
            return ""
        }
        return if (isGenericEpisodeTitle(value) || matchesLocalizedDisplayLabel(season, episodeNumber, value)) "" else value
    }

    @JvmStatic
    fun stripLeadingEpisodeMarker(season: String?, episodeNumber: String?, title: String?): String {
        val value = safe(title).trim()
        if (StringUtils.isBlank(value)) {
            return ""
        }
        val seasonDigits = normalizeDigitsToAscii(safe(season)).replace("\\D+".toRegex(), "")
        val episodeDigits = normalizeDigitsToAscii(safe(episodeNumber)).replace("\\D+".toRegex(), "")
        if (StringUtils.isBlank(episodeDigits)) {
            return value
        }
        var cleaned = value
        val separator = "\\s*[-:]*\\s*"
        if (StringUtils.isNotBlank(seasonDigits)) {
            cleaned = cleaned.replaceFirst("(?i)^\\s*season\\s*$seasonDigits\\s*(?:episode|ep|e)\\s*$episodeDigits\\b$separator".toRegex(), "")
            cleaned = cleaned.replaceFirst("(?i)^\\s*s\\s*$seasonDigits\\s*e\\s*$episodeDigits\\b$separator".toRegex(), "")
            cleaned = cleaned.replaceFirst("(?i)^\\s*$seasonDigits\\s*x\\s*$episodeDigits\\b$separator".toRegex(), "")
        }
        cleaned = cleaned.replaceFirst("(?i)^\\s*(?:episode|ep|e)\\s*$episodeDigits\\b$separator".toRegex(), "")
        return cleaned.trim()
    }

    @JvmStatic
    fun matchesLocalizedDisplayLabel(season: String?, episodeNumber: String?, title: String?): Boolean {
        val normalizedTitle = normalizeForComparison(title)
        if (StringUtils.isBlank(normalizedTitle)) {
            return true
        }
        val normalizedSeason = if (StringUtils.isBlank(season)) "1" else season!!
        val seasonLabel = I18n.formatSeasonLabel(normalizedSeason)
        val episodeLabel = I18n.formatEpisodeLabel(episodeNumber)
        val numericSeasonLabel = I18n.tr("autoSeasonNumber", I18n.formatNumber(normalizedSeason))
        val numericEpisodeLabel = I18n.tr("autoEpisodeNumber", I18n.formatNumber(if (StringUtils.isBlank(episodeNumber)) "-" else episodeNumber))

        for (candidate in listOf(
            "$seasonLabel - $episodeLabel",
            "$seasonLabel: $episodeLabel",
            "$seasonLabel $episodeLabel",
            "$seasonLabel  $episodeLabel",
            episodeLabel,
            "$seasonLabel - $numericEpisodeLabel",
            "$seasonLabel: $numericEpisodeLabel",
            "$seasonLabel $numericEpisodeLabel",
            "$numericSeasonLabel - $numericEpisodeLabel",
            "$numericSeasonLabel: $numericEpisodeLabel",
            "$numericSeasonLabel $numericEpisodeLabel",
            "$numericSeasonLabel - $episodeLabel",
            "$numericSeasonLabel: $episodeLabel",
            "$numericSeasonLabel $episodeLabel",
            numericEpisodeLabel
        )) {
            if (normalizeForComparison(candidate) == normalizedTitle) {
                return true
            }
        }
        val localizedEpisodeWord = localizedEpisodeWord(episodeLabel)
        if (StringUtils.isNotBlank(localizedEpisodeWord)) {
            val genericLocalizedEpisode = localizedEpisodeWord + " " + I18n.formatNumber(episodeNumber)
            if (normalizeForComparison(genericLocalizedEpisode) == normalizedTitle) {
                return true
            }
        }
        return false
    }

    @JvmStatic
    fun normalizeForComparison(value: String?): String {
        val normalized = Normalizer.normalize(normalizeDigitsToAscii(safe(value)), Normalizer.Form.NFKC)
            .lowercase(I18n.getCurrentLocale())
        return normalized.replace("[^\\p{L}\\p{N}]+".toRegex(), " ")
            .replace("\\s+".toRegex(), " ")
            .trim()
    }

    private fun normalizeDigitsToAscii(value: String): String {
        val out = StringBuilder(value.length)
        for (ch in value) {
            if (Character.isDigit(ch)) {
                val numeric = Character.getNumericValue(ch)
                if (numeric in 0..9) {
                    out.append(('0'.code + numeric).toChar())
                    continue
                }
            }
            out.append(ch)
        }
        return out.toString()
    }

    private fun safe(value: String?): String = value ?: ""

    private fun normalizeAsciiWords(value: String): String =
        normalizeDigitsToAscii(safe(value))
            .lowercase(Locale.ROOT)
            .replace('.', ' ')
            .replace(':', ' ')
            .replace('-', ' ')
            .trim()
            .replace("\\s+".toRegex(), " ")

    private fun matchesSimpleEpisodeNumber(value: String): Boolean =
        value.matches("^episode\\s+\\d+$".toRegex()) || value.matches("^ep\\s+\\d+$".toRegex()) || value.matches("^e\\d+$".toRegex())

    private fun matchesOrdinalSeasonEpisode(value: String): Boolean {
        val parts = value.split(" ")
        if (parts.size != 4 || parts[1] != "season" || !isEpisodeWord(parts[2])) {
            return false
        }
        return ORDINAL_URDU_PREFIXES.contains(parts[0]) && isDigits(parts[3])
    }

    private fun matchesSeasonEpisode(value: String): Boolean {
        val parts = value.split(" ")
        if (parts.size != 4 || parts[0] != "season" || !isEpisodeWord(parts[2])) {
            return false
        }
        return isDigits(parts[1]) && isDigits(parts[3])
    }

    private fun isEpisodeWord(value: String): Boolean = EPISODE_WORDS.contains(value)

    private fun isDigits(value: String): Boolean = StringUtils.isNotBlank(value) && value.chars().allMatch(Character::isDigit)

    private fun matchesLocalizedEpisodeWordTitle(normalizedTitle: String): Boolean {
        val localizedEpisodeWord = localizedEpisodeWord(I18n.formatEpisodeLabel("11"))
        if (StringUtils.isBlank(localizedEpisodeWord)) {
            return false
        }
        return normalizedTitle.matches(("^" + Pattern.quote(localizedEpisodeWord) + "\\s+\\d+$").toRegex())
    }

    private fun localizedEpisodeWord(episodeLabel: String?): String {
        val normalizedLabel = normalizeForComparison(episodeLabel)
        if (StringUtils.isBlank(normalizedLabel)) {
            return ""
        }
        val parts = normalizedLabel.split(" ")
        if (parts.size < 2) {
            return ""
        }
        val lastWord = parts.last()
        return if (lastWord.any { Character.isLetter(it) }) lastWord else ""
    }
}
